cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll

```diff
@@ -677,29 +664,28 @@ private predicate ssaFlowImpl(Node nodeFrom, Node nodeTo, boolean uncertain) {
  * Holds if `def` is the corresponding definition of
  * the SSA library's `definition`.
  */
-private predicate ssaDefinition(Def def, Definition definition) {
+private Definition ssaDefinition(Def def) {
   exists(IRBlock block, int i, SourceVariable sv |
     def.hasIndexInBlock(block, i, sv) and
-    definition.definesAt(sv, block, i)
+    result.definesAt(sv, block, i)
   )
 }
 
 /** Gets a node that represents the prior definition of `node`. */
-private Node getAPriorDefinition(Node node) {
-  exists(Def def, Definition definition, Definition inp, Def input |
-    defToNode(node, def, true) and
-    ssaDefinition(def, definition) and
-    uncertainWriteDefinitionInput(pragma[only_bind_into](definition), pragma[only_bind_into](inp)) and
-    ssaDefinition(input, inp) and
-    defToNode(result, input, _)
+private Node getAPriorDefinition(SsaDefOrUse defOrUse) {
+  exists(IRBlock bb, int i, SourceVariable sv, Definition def, DefOrUse defOrUse0 |
+    SsaCached::lastRefRedef(def, bb, i, ssaDefinition(defOrUse)) and
+    def.getSourceVariable() = sv and
+    defOrUse0.hasIndexInBlock(bb, i, sv) and
+    nodeToDefOrUse(result, defOrUse0, _)
```

Comment:
I think I need a voice-over for this change. This is part of 9ad4e8138c0db74f08b36cfc429c1b3d313bd752.

Reply:
Understandable! Consider this piece of code:
```cpp
int* p = indirect_source();
p[10] = 0;
sink(*p);
```
prior to this commit, we got flow in this example because we recognized that the definition `p[10] = 0` was a partial definition of `p`, and we'd use the shared SSA library's `uncertainWriteDefinitionInput` to find the previous definition (which in this case is `int* p = indirect_source()`), and thus we'd get flow from `int* p = indirect_source()` to `*p` because the SSA library concluded that there was flow from `p[10] = 0` to `sink(*p)` and `p[10] = 0` was partial. All good so far.

However, consider the failing testcase I added in https://github.com/github/codeql/commit/95222e37d67b0ee2a98ab3ac42ec67f83848f39f:
```cpp
intPointerSource(p);
p[10] = 0;
sink(*p); // $ MISSING: ast,ir
```
the old could would conclude that `p[10] = 0;` was a partial definition, and use `uncertainWriteDefinitionInput` to find the previous definition. However, `intPointerSource(p);` doesn't actually contain any SSA definitions. It does, however, contain an SSA _use_ of `*p`. So it turns out `uncertainWriteDefinitionInput` is the wrong tool for the job since it can only give us the previous definitions, but it won't have a result when the previous reference to the variable was a read instead of a def. So I swapped `uncertainWriteDefinitionInput` out with the right tool: `lastRefRedef`. That one gives us the (basic block, index) pair that last referenced the variable prior to a given definition, which is exactly what we need here (in this case the definition would be `p[10] = 0`, and the (basic block, index) pair would be the location of `p` in `intPointerSource(p)`).

Does that make sense?

Reply:
I'd be happy to do sync on it if it's more helpful.

Reply:
Makes sense.